linera-core/src/value_cache.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A least-recently used cache of [`HashedCertificateValue`]s.
+
+#[cfg(test)]
+#[path = "unit_tests/value_cache_tests.rs"]
+mod unit_tests;
+
+use std::{borrow::Cow, collections::BTreeSet, num::NonZeroUsize};
+
+use linera_base::crypto::CryptoHash;
+use linera_chain::data_types::{Certificate, HashedCertificateValue, LiteCertificate};
+use lru::LruCache;
+use tokio::sync::Mutex;
+#[cfg(with_metrics)]
+use {
+    linera_base::{prometheus_util, sync::Lazy},
+    prometheus::IntCounterVec,
+};
+
+use crate::worker::WorkerError;
+
+/// The default cache size.
+const DEFAULT_VALUE_CACHE_SIZE: usize = 1000;
+
+/// A counter metric for the number of cache hits in the [`CertificateValueCache`].
+#[cfg(with_metrics)]
+static CACHE_HIT_COUNT: Lazy<IntCounterVec> = Lazy::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "certificate_value_cache_hit",
+        "Cache hits in `CertificateValueCache`",
+        &[],
+    )
+    .expect("Counter creation should not fail")
+});
+
+/// A counter metric for the number of cache hits in the [`CertificateValueCache`].
+#[cfg(with_metrics)]
+static CACHE_MISS_COUNT: Lazy<IntCounterVec> = Lazy::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "certificate_value_cache_miss",
+        "Cache misses in `CertificateValueCache`",
+        &[],
+    )
+    .expect("Counter creation should not fail")
+});
+
+/// A least-recently used cache of [`HashedCertificateValue`]s.
+pub struct CertificateValueCache {
+    cache: Mutex<LruCache<CryptoHash, HashedCertificateValue>>,
+}
+
+impl Default for CertificateValueCache {
+    fn default() -> Self {
+        let size = NonZeroUsize::try_from(DEFAULT_VALUE_CACHE_SIZE)
+            .expect("Default cache size is larger than zero");
+
+        CertificateValueCache {
+            cache: Mutex::new(LruCache::new(size)),
+        }
+    }
+}
+
+impl CertificateValueCache {
+    /// Returns a [`BTreeSet`] of the hashes in the cache.
+    pub async fn keys(&self) -> BTreeSet<CryptoHash> {
+        self.cache
+            .lock()
+            .await
+            .iter()
+            .map(|(key, _)| *key)
+            .collect()
+    }
```

Comment:
I couldn't get rid of the allocation, because `Iterator`s are lazy and would have to at least take ownership of the `MutexGuard` in a hidden way, which I think wouldn't be ideal :sweat_smile: 

But I agree that it's best to avoid allocating a set for the _entire_ contents of the cache, so I managed to create a `subtract_cached_items_from` helper, which only allocates _after_ filtering, so it should be more efficient.

I also changed the `keys` to allow the caller to specify the type of the final `Collection`.